### ISSUE TITLE
Adds priority Class to Rancher helm chart

### ISSUE THREAD

**Issue by cbron** [rancher/rancher#37927]
We need to ensure Rancher pods are always scheduled ahead of other pods in the management cluster.

**Comment by cbron** [comment]
Original PR: https://github.com/rancher/rancher/pull/31830

**Comment by KevinJoiner** [comment]
The Current plan is to set the rancher pod and rancher feature charts priorityClassName via helm values stored in a config map.
This requires charts installed by Rancher to also have configurable priorityClassNames for their pods.

- [ ] https://github.com/rancher/webhook
- [ ] https://github.com/rancher/fleet
- [ ] https://github.com/rancher/aks-operator
- [ ] https://github.com/rancher/eks-operator
- [ ] https://github.com/rancher/gke-operator

**Comment by anupama2501** [comment]
On an upgrade from v2.6.5 >> v2.7-head b3c5e9f docker install of rancher, see the following warnings in docker logs scroll through every 5 minutes
```
2022/09/30 12:49:55 [WARNING] Failed to get rancher priorityClassName: failed to get rancher config: configmaps "rancher-config" not found
2022/09/30 12:49:55 [WARNING] Failed to get rancher priorityClassName: failed to get rancher config: configmaps "rancher-config" not found
2022/09/30 12:49:55 [WARNING] Failed to get rancher priorityClassName: failed to get rancher config: configmaps "rancher-config" not found
2022/09/30 12:54:55 [WARNING] Failed to get rancher priorityClassName: failed to get rancher config: configmaps "rancher-config" not found
2022/09/30 12:54:55 [WARNING] Failed to get rancher priorityClassName: failed to get rancher config: configmaps "rancher-config" not found
2022/09/30 12:54:55 [WARNING] Failed to get rancher priorityClassName: failed to get rancher config: configmaps "rancher-config" not found
2022/09/30 12:54:55 [WARNING] Failed to get rancher priorityClassName: failed to get rancher config: configmaps "rancher-config" not found
```

**Comment by KevinJoiner** [comment]
The above logs occur when Rancher is installed outside of helm. Since the required rancher-config is not present. This verbosity could be mitigated with an error not found check before the log statement.

**Comment by KevinJoiner** [comment]
## Root Cause
Pods critical to running Rancher did not use a priority class. This cause a cluster is limited resources to evict rancher pods before other non-critical pods.

## What was fixed, or what change has occurred
Add the ability to set the priority class for Rancher and its feature charts.

## Areas or cases that should be tested

We need to test that Rancher successfully applies PriorityClassNames to pods when deployed via helm.
The use of priorityClass during evictions and scheduling depends on Kubernetes and is not handled by Rancher.
### Verify Successful Docker Install
1. Start a new rancher server via Docker.
2. Wait 5 minutes.
3. Verify the logs do not contain the message `Failed to get rancher priorityClassName`
### Verify Helm install default priorityClass Propagation.
1. Preform a helm install of rancher
    ```  
    helm install rancher rancher-latest/rancher \                                                             
     --namespace cattle-system \
     --set hostname=<site_url>
     ```
2. Login set the server URL
3. Provision3 downstream cluster with GKE, AKD, and EKS
4. Verify the existence of the 'rancher-config' configMap
    ```
    kubectl get configmap -n cattle-system rancher-config -o yaml
    ```
5. Verify that rancher-config has the field `priorityClassName` set to `rancher-critical`
6. Verify the existence of the rancher `rancher-critical' priorityClass
   ```
   kubectl get priorityclass rancher-critical -o yaml
   ```
7. Verify that `rancher-critical` has the field `value` set to `1000000000`
8. Verify that the rancher deployment has `priorityClassName` set to `rancher-critical`
    ```
     kubectl get deployment -n cattle-system rancher -o yaml
    ```
9. Verify that the webhook deployment has `priorityClassName` set to `rancher-critical`
    ```
     kubectl get deployment -n cattle-system rancher-webhook -o yaml
    ``` 
10. Verify that the fleet-controller deployment has `priorityClassName` set to `rancher-critical`
    ```
     kubectl get deployment -n cattle-fleet-system fleet-controller -o yaml
    ``` 
11. Verify that the gitjob deployment has `priorityClassName` set to `rancher-critical`
    ```
     kubectl get deployment -n cattle-fleet-system gitjob -o yaml
    ``` 
11. Verify that the gke-config-operator deployment has `priorityClassName` set to `rancher-critical`
    ```
     kubectl get deployment -n cattle-system gke-config-operator -o yaml
    ``` 
12. Verify that the aks-config-operator deployment has `priorityClassName` set to `rancher-critical`
    ```
    kubectl get deployment -n cattle-system aks-config-operator -o yaml
    ``` 
13. Verify that the eks-config-operator deployment has `priorityClassName` set to `rancher-critical`
    ```
     kubectl get deployment -n cattle-system eks-config-operator -o yaml
    ``` 
### Verify Helm install custom priorityClass Propagation.
1. Preform a helm install of rancher
    ```  
    helm install rancher rancher-latest/rancher \                                                             
     --namespace cattle-system \
     --set hostname=<site_url> \
     --set priorityClassName="system-node-critical"
     ```
2. Login set the server URL
3. Verify the existence of the 'rancher-config' configMap
    ```
    kubectl get configmap -n cattle-system rancher-config -o yaml
    ```
4. Verify that rancher-config has the field `priorityClassName` set to `system-node-critical`
5. Verify that the rancher deployment has `priorityClassName` set to `system-node-critical`
    ```
     kubectl get deployment -n cattle-system rancher -o yaml
    ```
6. Verify that the webhook deployment has `priorityClassName` set to `system-node-critical`
    ```
     kubectl get deployment -n cattle-system rancher-webhook -o yaml
    ``` 
## What areas could experience regressions
Rancher upgrade via helm has priorityClass set.

## Are the repro steps accurate/minimal?
- Install Rancher server via helm.
- Verify rancher deployment does not have a priority class set.

**Comment by dasarinaidu** [comment]
Validations completed and looks good to me. 

Validated below test cases as part of this card validations.

<img width="1130" alt="image" src="https://user-images.githubusercontent.com/22034884/196499837-40c85386-058e-4486-b661-76fd788c9b2c.png">

**Comment by dasarinaidu** [comment]
Validations completed and looks good to me.

**Rancher versions:**
v2.7-9f1e0435bb45e104ba897c5c329e1f23735843d9-head
upgrade version : v2.6.8 to v2.7.0-rc3

Validated below test cases as part of this card validations.

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/sdasari/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/sdasari/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">



Test   Scenario | Test Cases | Check point | Status |  
-- | -- | -- | -- | --
Docker Install rancher | Local   cluster | Verify   the logs do not contain the message Failed to get rancher priorityClassName | Pass |  
  | Create   a cluster | RKE/EKS/AKS | Pass |  
Helm install rancher | Create   AKE, GKE & EKS clusters     On Local cluster:kubectl | kubectl   get configmap -n cattle-system rancher-config -o yaml | Pass |  
  |   | kubectl   get priorityclass rancher-critical -o yaml | Pass |  
  |   | kubectl get deployment -n cattle-system   rancher -o yaml | Pass |  
  |   | kubectl get deployment -n cattle-system   rancher-webhook -o yaml | Pass |  
  |   | kubectl get deployment -n   cattle-fleet-system fleet-controller -o yaml | Pass |  
  |   | kubectl get deployment -n   cattle-fleet-system gitjob -o yaml | Pass |  
  |   | kubectl get deployment -n cattle-system   gke-config-operator -o yaml | Pass |  
  |   | kubectl   get deployment -n cattle-system aks-config-operator -o yaml | Pass |  
  |   | kubectl get deployment -n cattle-system   eks-config-operator -o yaml | Pass |  
Helm install rancher | --set   priorityClassName="system-node-critical" | kubectl   get configmap -n cattle-system rancher-config -o yaml | Pass |  
  |   | kubectl get deployment -n cattle-system   rancher-webhook -o yaml | Pass |  
  | After   upgrade : Added few clusters | Make   sure the cluster creation working good | Pass |  
Helm Upgrade rancher | Install   v2.6.8     Create AKE, GKE & EKS clusters     Upgrade rancher to v2.7.0-rc3 | kubectl   get configmap -n cattle-system rancher-config -o yaml | Pass |  
  |   | kubectl   get priorityclass rancher-critical -o yaml | Pass |  
  |   | kubectl get deployment -n cattle-system   rancher -o yaml | Pass |  
  |   | kubectl get deployment -n cattle-system   rancher-webhook -o yaml | Pass |  
  |   | kubectl get deployment -n   cattle-fleet-system fleet-controller -o yaml | Pass |  
  |   | kubectl get deployment -n   cattle-fleet-system gitjob -o yaml | Pass |  
  |   | kubectl get deployment -n cattle-system   gke-config-operator -o yaml | Pass |  
  |   | kubectl   get deployment -n cattle-system aks-config-operator -o yaml | Pass |  
  |   | kubectl get deployment -n cattle-system   eks-config-operator -o yaml | Pass |  
Helm Upgrade rancher | --set   priorityClassName="system-node-critical" | kubectl   get configmap -n cattle-system rancher-config -o yaml | Pass |  
  |   | kubectl get deployment -n cattle-system   rancher-webhook -o yaml | Pass |  
  | After   upgrade : Added few clusters | Make   sure the cluster creation working good | Pass |  
Update flag | Update   the flag from 'rancher-critical' to 'system-node-critical' | On   v2.7.0-rc3 : Check for the flag update (local->deployments, all name   spaces-.Edit config->pods->resources->priprity class name) | Pass |  
  | Delete   the flag   'system-node-critical' from   deployments | On   v2.7.0-rc3 : Check for the flag update (local->deployments, all name   spaces-.Edit config->pods->resources->priprity class name) | Pass | From   rancher UI I deleted the priorityClassName    flag (from system-node-critical to blank) on deployments and it   deletes only on deployments. The flag stays the same on webhook and EKS   clusters.
  |   | Make   sure the flag is not removed from eks cluster | Pass |  
  | Delete   the flag   'system-node-critical' from   webhook | On   v2.7.0-rc3 : Check for the flag update (local->deployments, all name   spaces-.Edit config->pods->resources->priprity class name) | Pass |  



</body>

</html>

**Comment by KevinJoiner** [comment]
# Release Notes
## Synopsis of issue
Pods critical to running Rancher did not use a priority class. This could cause a cluster with limited resources to evict rancher pods before other non critical pods.
## Resolved (or not):
Resolved by adding a configurable priorityClass to the Rancher pod required feature charts.
## Status: 
resolved in v2.7.0
## Versions: 
< v2.7.0
